GitHub: Add "Clone and Open" action

daquinoaldo opened the pull request:
## Description

Add a `Clone and Open` command in the `Search Repositories` results, with shortcut `⌘⌥C`.
The command clones the repository in a configurable path (by default, `~/GitHub`). Then, it opens the cloned folder using a configurable IDE (by default, VS Code).
The clone is skipped if the repository already exists. When cloning, it uses the `git` binary (pre-installed on MacOS) and HTTPS with the extension's OAuth token to avoid any user configuration requirement.

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

daquinoaldo commented on extensions/github/src/helpers/repository.ts:
```diff
@@ -34,15 +34,61 @@ export const WEB_IDES = [
+
+  if (!existsSync(clonePath.replace("~", homedir()))) {
+    const gitHubToken = getGitHubToken();
+    const cloneUrl = `https://oauth:${gitHubToken}@github.com/${repository.nameWithOwner}`;
```
I tried many commands, but in the end, I found this the most robust.

- SSH should be preferred but requires the user to set it up, which may lead to unpleasant failures. Also, when the user uses [1Password CLI](https://developer.1password.com/docs/ssh/) or other password manager integrations to protect their SSH keys, the SSH clone won't work because the user doesn't get the authentication prompt, and the command fails as unauthenticated.
- The `gh` CLI is an alternative method but forces the user to install the CLI. We could install it for them, but I don't like forcing such dependency when it's not needed and doesn't provide much additional value. In any case, it also requires more complex error-handling logic, making this command unnecessarily cumbersome. Finally, if the user protects the `gh` CLI token with password managers (e.g., [1Password](https://developer.1password.com/docs/cli/shell-plugins/github/)), it won't work for the same reason as SSH.

We already store and use a GitHub token, either user-generated or from OAuth, and it already has the permissions required for the clone, so it's pointless to ask the user for further authentications.

daquinoaldo commented on extensions/github/src/components/RepositoryActions.tsx:
```diff
@@ -107,6 +107,13 @@ export default function RepositoryActions({ repository, mutateList, onVisit }: R
+
         <Action.OpenInBrowser
           icon="vscode.svg"
           title="Clone in VSCode"
```
I know there was already this command available. However, I think it's very limited and nowhere near as powerful as the new one.
It allows you to open VS Code with the clone prompt but doesn't let you open an already-cloned repository from the search results. That's why I added the Clone and Open: you clone once and open as many times as you want.
In my opinion, we could deprecate the Clone in VSCode command. However, it would be a breaking change, and I don't know what's the best practice in this case. Personally, I don't have any preference in this regard, and we could leave it there.

thomaslombart replied:
I think it's fine having them both. The VS Code action always prompt you to select the repository destination so it's slightly different.

thomaslombart commented on extensions/github/package.json:
```diff
@@ -193,6 +194,22 @@
     }
   ],
   "preferences": [
+    {
+      "name": "application",
+      "type": "appPicker",
+      "required": false,
+      "title": "IDE Application",
+      "description": "Will be used to open the cloned repository.",
+      "placeholder": "/Applications/Visual Studio Code.app"
+    },
+    {
+      "name": "baseClonePath",
+      "type": "directory",
+      "required": false,
+      "title": "Clone Path",
+      "description": "Path in which the repositories will be cloned.",
+      "placeholder": "~/GitHub"
+    },
```
You can use the `default` property over here instead.

thomaslombart commented on extensions/github/package.json:
```diff
@@ -193,6 +194,22 @@
     }
   ],
   "preferences": [
+    {
```
Could you move both preferences in the `Search Repositories` command preferences? That way it doesn't pollute the global preferences for these not using it.

thomaslombart reviewed (commented):
Great work!

daquinoaldo commented:
Thanks for the review, @thomaslombart! It's my first contribution, so I'm not yet familiar with the SDK. 😄 I should have addressed all the comments, can you have a second look?

thomaslombart commented on extensions/github/src/helpers/repository.ts:
```diff
@@ -34,15 +34,59 @@ export const WEB_IDES = [
+    const cloneCommand = `git clone ${cloneUrl} ${clonePath}`;
+
+    try {
+      execSync(cloneCommand);
```
Missed that one, but you probably want to show a success toast when it's done. Otherwise, you'll get the animated toast infinitely.

daquinoaldo replied:
Updated, thanks.

thomaslombart commented:
Great! 🚀
